File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:

        ans = []

        word_len = len(words[0])
        word_count = len(words)

        need = {}

        for word in words:
            need[word] = need.get(word, 0) + 1

        for offset in range(word_len):

            left = offset
            count = 0

            window = {}

            right = offset

            while right + word_len <= len(s):

                word = s[right:right + word_len]

                # Word is not required
                if word not in need:

                    window.clear()
                    count = 0
                    left = right + word_len

                    right += word_len
                    continue

                # Add word
                window[word] = window.get(word, 0) + 1
                count += 1

                # Too many copies
                while window[word] > need[word]:

                    left_word = s[left:left + word_len]

                    window[left_word] -= 1

                    left += word_len
                    count -= 1

                # All words found
                if count == word_count:

                    ans.append(left)

                    left_word = s[left:left + word_len]

                    window[left_word] -= 1

                    left += word_len
                    count -= 1

                right += word_len

        return ans

File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.java
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {

        List<Integer> ans = new ArrayList<>();

        int wordLen = words[0].length();
        int wordCount = words.length;

        HashMap<String, Integer> need = new HashMap<>();

        for (String word : words) {
            need.put(word, need.getOrDefault(word, 0) + 1);
        }

        for (int offset = 0; offset < wordLen; offset++) {

            int left = offset;
            int count = 0;

            HashMap<String, Integer> window = new HashMap<>();

            for (int right = offset;
                 right + wordLen <= s.length();
                 right += wordLen) {

                String word = s.substring(right, right + wordLen);

                // Word is not required
                if (!need.containsKey(word)) {

                    window.clear();
                    count = 0;
                    left = right + wordLen;

                    continue;
                }

                // Add word
                window.put(
                    word,
                    window.getOrDefault(word, 0) + 1
                );

                count++;

                // Too many copies
                while (window.get(word) > need.get(word)) {

                    String leftWord =
                        s.substring(left, left + wordLen);

                    window.put(
                        leftWord,
                        window.get(leftWord) - 1
                    );

                    left += wordLen;
                    count--;
                }

                // All words found
                if (count == wordCount) {

                    ans.add(left);

                    String leftWord =
                        s.substring(left, left + wordLen);

                    window.put(
                        leftWord,
                        window.get(leftWord) - 1
                    );

                    left += wordLen;
                    count--;
                }
            }
        }

        return ans;
    }
}